include/envoy/http/header_map.h

```diff
@@ -637,12 +631,90 @@ class HeaderMap {
 
 using HeaderMapPtr = std::unique_ptr<HeaderMap>;
 
+/**
+ * Registry for custom headers. Headers can be registered multiple times in independent
+ * compilation units and will still point to the same slot. Headers are registered independently
+ * for each concrete header map type and do not overlap. Handles are strongly typed and do not
+ * allow mixing.
+ */
+template <class T> class CustomInlineHeaderRegistry {
+public:
+  using RegistrationMap = std::map<LowerCaseString, size_t>;
+  // Phantom is used here to force the compiler to verify that handles are not mixed between
+  // concrete header map types.
+  using Handle = Phantom<RegistrationMap::const_iterator, T>;
+
+  /**
+   * Register an inline header and return a handle for use in inline header calls. Must be called
+   * prior to finalize().
+   */
+  static Handle registerInlineHeader(const LowerCaseString& header_name) {
+    static size_t inline_header_index = 0;
+
+    ASSERT(!mutableFinalized());
```

Comment:
it is surprising that you are asserting we are not finalized here, and then in the next line calling an accessor that asserts it is finalized. I don't see the intervening call to finalize().

Reply:
The accessor doesn't assert finalized, it just gets the mutable variable.